include/kokkos-utils/callbacks/ConjunctionMatcher.hpp

```diff
@@ -0,0 +1,32 @@
+#ifndef KOKKOS_UTILS_CALLBACKS_CONJUNCTIONMATCHER_HPP
+#define KOKKOS_UTILS_CALLBACKS_CONJUNCTIONMATCHER_HPP
+
+#include "kokkos-utils/callbacks/Events.hpp"
+#include "kokkos-utils/callbacks/Matcher.hpp"
+
+namespace Kokkos::utils::callbacks
+{
+//! Conjunction of matchers that is @c true only if all @ref matchers agree.
+template <typename... Matchers> requires (sizeof...(Matchers) > 1)
+struct ConjunctionMatcher
+{
+    template <typename... Args> requires (sizeof...(Args) == sizeof...(Matchers))
+    explicit ConjunctionMatcher(Args&&... args) : matchers(std::forward<Args>(args)...) {}
+
+    template <Event EventType> requires (MatcherFor<Matchers, EventType> && ...)
+    bool operator()(const EventType& event)
+    {
+        return std::apply([&event] (Matchers&... matchers) {
+                return (matchers(event) && ...);
+        }, matchers);
```

Comment:
We could avoid evaluating the second sub-matcher if the first one is already `false`.

Reply:
Because `std::conjunction` is short-circuiting, so should this one.

And we'd have another helper called `AndMatcher` that would evaluate all sub-matchers and perform `AND`.

@maartenarnst What do you think ?

Reply:
It seems it's already all good [as stated in this order of evaluation rule](https://en.cppreference.com/w/cpp/language/eval_order):
> 6) The first (left) operand of the built-in [logical](https://en.cppreference.com/w/cpp/language/operator_logical) AND operator &&, the built-in logical OR operator || and the built-in [comma operator](https://en.cppreference.com/w/cpp/language/operator_other#Built-in_comma_operator) , is sequenced before the second (right) operand.